source/distributed/shared_tria.cc

```diff
@@ -319,6 +323,50 @@ namespace parallel
 
 
 
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::set_subdomain_mask(
+      const SubdomainMask new_mask)
+    {
+      if (allow_artificial_cells && new_mask != subdomain_mask)
+        switch (new_mask)
+          {
+            case SubdomainMask::true_ids:
+              // Save the current set of subdomain IDs, and set subdomain IDs
+              // to the "true" owner of each cell.
+              Assert(subdomain_mask == SubdomainMask::default_ids,
+                     ExcInternalError());
+              default_subdomain_ids.resize(this->n_active_cells());
+              for (const auto &cell : this->active_cell_iterators())
+                {
+                  const unsigned int index     = cell->active_cell_index();
+                  default_subdomain_ids[index] = cell->subdomain_id();
+                  cell->set_subdomain_id(true_subdomain_ids_of_cells[index]);
+                }
+              break;
+
+            case SubdomainMask::default_ids:
+              // Undo the subdomain modification.
+              Assert(subdomain_mask == SubdomainMask::true_ids,
+                     ExcInternalError());
+              for (const auto &cell : this->active_cell_iterators())
+                {
+                  const unsigned int index = cell->active_cell_index();
+                  cell->set_subdomain_id(default_subdomain_ids[index]);
+                }
+              std::vector<types::subdomain_id>().swap(default_subdomain_ids);
+              break;
+
+            default:
+              Assert(false, ExcNotImplemented());
+              break;
+          }
+
+      subdomain_mask = new_mask;
+    }
+
+
+
```

Comment:
What happens if you have set the mask, and then do mesh refinement, and then reset it to `default_ids`. Isn't the array the wrong size then? I think you probably want to reset to default upon mesh refinement, and `clear()` the array.